pkg/dhcp6client/dhcp6client.go

```diff
@@ -106,38 +108,40 @@ func (p *buffer) remaining() []byte {
 	return p.consume(len(p.buf))
 }
 
+// If the client fails to receive a valid DHCPv6 packet via socket,
+// it keeps listening until the time is out.
 func (c *Client) ReadReply() (*dhcp6.Packet, error) {
+	tick := time.NewTicker(c.timeout)
+	start := time.Now()
+
 	var err error
-	for i := 0; i < 5; i++ {
-		// set a random timeout within range [c.timeout, 2*c.timeout)
-		// so as to prevent swamping a DHCP server.
-		c.connection.SetReadTimeout(c.timeout + time.Duration(rand.Intn(int(c.timeout))))
-		pb := make([]byte, 1500)
-		var n int
-		n, err = c.connection.ReadFrom(pb)
-		if err != nil {
-			// Return error if read time of socket is due,
-			// so that request can be resent instantly
-			if err, ok := err.(net.Error); ok && err.Timeout() {
-				return nil, err
+	for {
+		select {
+		case <-tick.C:
+			return nil, err
+		default:
+			c.connection.SetReadTimeout(c.timeout - time.Since(start))
```

Comment:
I don't think you really need the ticker. You could just check whether `c.timeout - time.Since(start)` is 0 or less than 0.

```go
start := time.Now()

for {
  timeout := c.timeout - time.Since(start)
  if timeout <= 0 {
    return nil, ErrRetry
  }
  c.connection.SetReadTimeout(timeout)
  ...
}
```